### PR TITLE
Fix thread pool completed work item counter to remove finalized nodes

### PR DESCRIPTION
**Pull request by kouvel**
- New nodes are being added to the set but finalized ones were not being removed
- The node was also not made finalizable, fixed by adding a finalization helper with the appropriate lifetime for each node

### CODE REVIEW

**Review by davidwrighton — CHANGES_REQUESTED** [review]
I don't think this change addresses the problem at all. Since there is a reference path from PortableThreadPool through the ThreadInt64PersistentCounter to each ThreadLocalNode, I don't see any way for the finalizer on `ThreadLocalNode` to ever execute. I recently learned about `ThreadLocal<T>` and I think creative use of that api would likely be a more effective way to fix this issue. In particular, remove the finalizer from `ThreadLocalNode`, and use `ThreadLocal<T>` to setup a finalizer that when a thread exits will call a Dispose function on `ThreadLocalNode`.

**Comment by kouvel** [comment]
Yep missed that, will need a better fix

File: src/libraries/System.Private.CoreLib/src/System/Threading/ThreadInt64PersistentCounter.cs [hunk 1]
```diff
@@ -65,7 +75,7 @@ public long Count
             }
         }
 
-        private sealed class ThreadLocalNode
+        private sealed class ThreadLocalNode : IDisposable
```
Comment (davidwrighton) [inline src/libraries/System.Private.CoreLib/src/System/Threading/ThreadInt64PersistentCounter.cs:78]: IDisposable here isn't necessary.

**Review by davidwrighton — APPROVED** [review]
This should work nicely.

File: src/libraries/System.Private.CoreLib/src/System/Threading/ThreadInt64PersistentCounter.cs [hunk 2]
```diff
@@ -26,6 +29,13 @@ public object CreateThreadLocalCountObject()
         {
             var node = new ThreadLocalNode(this);
 
+            List<ThreadLocalNodeFinalizationHelper>? nodeFinalizationHelpers = t_nodeFinalizationHelpers;
+            if (nodeFinalizationHelpers == null)
+            {
+                t_nodeFinalizationHelpers = nodeFinalizationHelpers = new List<ThreadLocalNodeFinalizationHelper>(1);
+            }
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Threading/ThreadInt64PersistentCounter.cs:36]:
```suggestion
            List<ThreadLocalNodeFinalizationHelper>? nodeFinalizationHelpers = t_nodeFinalizationHelpers ??= new List<ThreadLocalNodeFinalizationHelper>(1);
```

**Review by stephentoub — APPROVED** [review]

**Comment by danmoseley** [comment]
Failure was the xunit crash. 😢

**Pull request merged**